system/Validation/Validation.php

```diff
@@ -193,6 +193,29 @@ public function check($value, string $rule, array $errors = []): bool
 	 */
 	protected function processRules(string $field, string $label = null, $value, $rules = null, array $data)
 	{
+		// If the if_exist rule is defined...
+		if (in_array('if_exist', $rules))
+		{
+			// and the current field does not exists in the input data
+			// we can return true. Ignoring all other rules to this field.
+			if (! array_key_exists($field, $data))
+			{
+				return true;
+			}
+			// Otherwise remove the if_exist rule and continue the process
+			$rules = array_diff($rules, ['if_exist']);
+		}
+
+		if (in_array('permit_empty', $rules))
+		{
+			if (! in_array('required', $rules) && (is_array($value) ? empty($value) : (trim($value) === '')))
```

Comment:
With this condition the following input values will not be considered "empty" values.

- "0" (0 as a string)
- 0 (0 as an integer)
- 0.0 (0 as a float)

Because of the `trim()` returns "0" and "0.0" strings.

My doubt about is that this condition is used in the [required rule](https://github.com/bcit-ci/CodeIgniter4/blob/8da268b8ea13ed94b279de5375f10632dc6fa0b5/system/Validation/Rules.php#L280) and in the [user guide](https://github.com/bcit-ci/CodeIgniter4/blame/develop/user_guide_src/source/libraries/validation.rst#L638) we have the description "Fails if the field is empty.". 

Maybe it must be updated to "Fails if the field is an empty array, empty string, null or false."? 

Since the numeric empty's are not considered empty's, contrary of [empty()](http://php.net/manual/en/function.empty.php#refsect1-function.empty-returnvalues). Or use `empty()` for all data types and not compare if the non-arrays are identical to an empty string as it is being done `(trim($str) !== '')`.

It will make more sense to update the description, I believe.

Please, talk about.

Reply:
I think we should clarify in the docs. You're correct that numeric values should not be considered "empty" but actual values, since it's a fairly common thing to want a number `0` as input.